src/Input/Input.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file was originally part of the league/commonmark package.
+ *
+ * (c) Colin O'Dell <colinodell@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace UnicornFail\Emoji\Input;
+
+use UnicornFail\Emoji\Exception\UnexpectedEncodingException;
+
+class Input implements InputInterface
+{
+    /**
+     * @var ?iterable<int, string>
+     *
+     * @psalm-readonly-allow-private-mutation
+     */
+    private $lines;
```

Comment:
I don't think you necessarily need the `src/Input` stuff from commonmark.  We need it there because implementing the spec requires us to consider **each line** as a whole since the parsing behavior might change based on what the previous/next line contains.

In this case, I don't think we necessarily need to consider lines since we're primarily working at the character-level - I don't think `\n` has any special meaning, at least not one that would be different from other whitespace characters.

Reply:
I think this is needed, if only from a performance level (memory usage). I could be mistaken, but it seems like it would make more sense if `preg_split` worked on individual lines, rather than the document as a whole.